common/main/java/com/couchbase/lite/internal/replicator/AbstractCBLWebSocket.java

```diff
@@ -297,8 +331,16 @@ protected AbstractCBLWebSocket(
 
     @Override
     protected void openSocket() {
+        OkHttpClient okHttpClient = httpClient;
+        if (pingInterval > 0) {
```

Comment:
I worry that this is going to conflict with our upcoming feature to make the Core heartbeat configurable by the CBL consumer.  It effectively means we'd now have two ways to set the same feature.  Most of the time when the network is truly down, the TCP connection will also die which will close the web socket connection.  This setting is really only useful when the TCP connection is still alive but the other side is not responding for whatever reason.  I'd be interested to hear about scenarios like this, that would not be solved by a configurable core heartbeat.  Do you know of any?

Reply:
Heartbeat is necessary. As couchbase lite is for mobile devices, any bad point between client and server will make it happen. Especially in our situation, each of our android device doesn't access to internet directly, but through another linux box that communicate with internet by 4G. So, when network is bad, you know... The mobile device can't detect it directly, maybe heartbeat is the only way.

And it is ok for us if core does heartbeats. That makes no difference. What we want include several points:
1. Configurable interval in java codes
2. Fail in time when network down is detected

According to the version I read, there is heartbeat in the core, but only for framing mode. Will it be changed?

Reply:
Interesting question.  The fact that Java is using its own framing is something I was actually unaware of.  Apple and .NET both using framing.  If I ponder out loud in this comment, the whole existence of framing is basically a choice that the platform side makes between "I want Core to format my messages for me so that I can just send them over a raw socket" or "I want Core to give me the message payload and I will format / send it."  Extending that, it means that it is not possible for Core to send a heartbeat if framing is not being used because it is in the dark about how to send it (or if it needs to be sent at all).  Side note: The current other choice besides "no framing" is not "framing", but "web socket client framing" which is why Core can confidently send the heartbeat on its own.  So in light of that, you are correct that Java must be the master of its own heartbeat.  That being said, I am going to assume that this difference is going to be hidden behind the new API that we will offer in the next 2.8.x release to make both the retry logic and heartbeat configurable, rendering the public facing part of this PR moot and conflicting (the implementation, however, can remain).

Some optional background information that hopefully explains why we are introducing this API change, and why it will also satisfy your requirement, is that we have a large number of users who put Sync Gateway between a network load balancer which tends to cut off connections it views as "idle."  This is bad for continuous replication, which spends large amounts of time idling.  Currently we have the heartbeat set at a hard coded 5 minutes, but very often load balancers cut off connections before that and so we are going to allow the consumer to configure the heartbeat interval.  As it stands now, the purpose of the heartbeat is twofold:  It both makes sure that the other side does not hang up, and it checks the health of the connection.  If a reply is not received in a timely manner (10 seconds), the connection is failed and replication stops (I forget if this is considered recoverable or not).  

All of this is unrelated in a platform that does not use framing, though, so Java is free to do as it wishes in terms of heartbeats as long as it conforms to the API we will provide (out of your control for now) and accomplishes the same two goals, the second of which I believe is also your goal.  My recommendation is to leave the replicator configuration changes out of the PR (since they will just be overwritten later when the new API comes) and keep the implementation.  @bmeike Does this sound reasonable?

Reply:
@borrrden You mean, for java and android, enabling heartbeat this way is ok, but the user interface should comply with your coming version, right? So I should revert some code for the Configuration class, maybe CBLWebSocket remains. Is that ok?

Reply:
Yes that is what my recommendation is.  In general the changes of merging something into the public API space from outside users is low because we need to weigh them against the need to bring the same features to other platforms besides (in this case) Java.  In this particular case though, we are already planning the same change that you desire (though your PR has helped us realize this mistake in Java that otherwise would have continued to go unnoticed, I think) so the functionality is definitely welcome, with the API to be introduced later.

That being said, @bmeike must sign off on this as the owner of this repo.  This is only my personal recommendation as a sort of "outside counsel" (I work with LiteCore and I basically came up here to explain why things are the way they are at the core level, and my recommendation might have been an overreach).